src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -8,16 +8,129 @@
+        /// <see cref="TryComplete"/> is called.  This transition will not happen synchronously.
+        /// 
+        /// This Task will not complete until it has completed all existing values returned
+        /// from <see cref="DequeueAsync"/>.
+        /// </summary>
+        public Task WhenCompletedTask
+            => _whenCompleted.Task;
+
+        /// <summary>
+        /// Gets a value indicating whether the queue has completed.
+        /// </summary>
+        public bool IsCompleted
+            => _whenCompleted.Task.IsCompleted;
+
+        /// <summary>
+        /// The number of unconsumed elements in the queue.
+        /// </summary>
+        public int Count => _channel.Reader.Count;
+
+        /// <summary>
+        /// Adds an element to the tail of the queue.  This method will throw if the queue 
+        /// is completed.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is already completed.</exception>
+        /// <param name="value">The value to add.</param>
+        public void Enqueue(TElement value)
+        {
+            if (!TryEnqueue(value))
+                throw new InvalidOperationException($"Cannot call {nameof(Enqueue)} when the queue is already completed.");
+        }
+
+        /// <summary>
+        /// Tries to add an element to the tail of the queue.  This method will return false if the queue
+        /// is completed.
+        /// </summary>
+        /// <param name="value">The value to add.</param>
+        public bool TryEnqueue(TElement value)
+            => _channel.Writer.TryWrite(value);
+
+        /// <summary>
+        /// Attempts to dequeue an existing item and return whether or not it was available.
+        /// </summary>
+        public bool TryDequeue(out TElement d)
+            => _channel.Reader.TryRead(out d);
+
+        /// <summary>
+        /// Same operation as <see cref="AsyncQueue{TElement}.Complete"/> except it will not
+        /// throw if the queue is already completed.
+        /// </summary>
+        /// <returns>Whether or not the operation succeeded.</returns>
+        public bool TryComplete()
+        {
+            var result = _channel.Writer.TryComplete();
+            _whenCompleted.TrySetResult(true);
+            return result;
+        }
+
+        /// <summary>
+        /// Signals that no further elements will be enqueued.  All outstanding and future
+        /// Dequeue Task will be cancelled.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is already completed.</exception>
+        public void Complete()
+        {
+            if (!TryComplete())
+            {
+                throw new InvalidOperationException($"Cannot call {nameof(Complete)} when the queue is already completed.");
+            }
+        }
+
+        public void PromiseNotToEnqueue()
+            => TryComplete();
```

Comment:
this method seems superrrrr weird.  it's only used in *One* location in the entire codebase in code that looks like this:

```
            // Signal the end of compilation.
            EventQueue.TryEnqueue(new CompilationCompletedEvent(this));
            EventQueue.PromiseNotToEnqueue();
            EventQueue.TryComplete();
```

since we're claling TryComplete right after it, i see no purpose in this.  would like to remove, but i want a binary drop-in for testing.